File: src/main/java/pro/tuscan/domain/profiles/LatestProfileService.java
package pro.tuscan.domain.profiles;

import pro.tuscan.adapter.api.request.LatestProfileRequest;
import pro.tuscan.supplier.InstantSupplier;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class LatestProfileService {

    private final InstantSupplier instantSupplier;
    private final LatestProfileRepository latestProfileRepository;

    public LatestProfileService(InstantSupplier instantSupplier,
                                LatestProfileRepository latestProfileRepository) {

        this.instantSupplier = instantSupplier;
        this.latestProfileRepository = latestProfileRepository;
    }

    public Flux<LatestProfile> findLatestProfiles() {
        return latestProfileRepository.findTop6ByOrderByCreatedOnDesc();
    }

    public Mono<LatestProfile> save(LatestProfileRequest request) {
        return Mono.just(request)
                .map(r -> LatestProfileMapper.Companion.mapAndUpdate(r, instantSupplier.get()))
                .flatMap(latestProfileRepository::save);
    }
}

File: src/main/java/pro/tuscan/domain/profiles/LatestProfileService.kt
package pro.tuscan.domain.profiles

import org.springframework.stereotype.Service
import pro.tuscan.adapter.api.LatestProfileRequest
import pro.tuscan.supplier.InstantSupplier
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

@Service
class LatestProfileService(private val instantSupplier: InstantSupplier,
                           private val latestProfileRepository: LatestProfileRepository) {

    fun findLatestProfiles(): Flux<LatestProfile> =
            latestProfileRepository.findTop6ByOrderByCreatedOnDesc()

    fun save(request: LatestProfileRequest): Mono<LatestProfile> =
            Mono.just(request)
                    .map { req -> mapAndUpdate(req, instantSupplier.get()) }
                    .flatMap { entity -> latestProfileRepository.save(entity) }

    private fun mapAndUpdate(request: LatestProfileRequest, instant: Instant): LatestProfile =
            request.let {
                LatestProfile(
                        nickname = it.nickname,
                        avatarUrl = it.avatarUrl,
                        level = it.level,
                        elo = it.elo,
                        kdRatio = it.kdRatio,
                        createdOn = instant)
            }
}
